doc/convickt.1
.TH CONVICKT 1
.\" NAME should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)
.\"
.\" Created by Alex Smith; licensed under GNU GPL v2 or later.
.SH NAME
convickt \- convert INTERCAL files between formats
.SH SYNOPSIS
.B convickt
.I "incharset outcharset [paddingrule [arrayname]]"
.SH "DESCRIPTION"
This manual page describes the
.BR convickt
command.
Fuller documentation is available via the command
.B info convickt
.
.PP
.B convickt
converts INTERCAL programs from one character set to another; it takes
input from standard input and sends its output to standard output.
.SH OPTIONS
For
.I incharset
and
.I outcharset
.TP
.B atari
Interpret the input as being ASCII-7 Atari syntax INTERCAL, as is used
by the Atari INTERCAL-72 compiler, J-INTERCAL, and C-INTERCAL default
syntax, or output in that syntax.
.TP
.B baudot
Interpret the input as being in CLC-INTERCAL's extended Baudot
syntax, or output in that syntax.
.TP
.B ebcdic
Interpret the input as being in the CLC-INTERCAL dialect of EBCDIC, or
output in that syntax.
.TP
.B latin1
Interpret the input as being Latin-1 Princeton syntax INTERCAL, as is
used by default by CLC-INTERCAL and also readable by C-INTERCAL with
the
.B \-X
switch, or output in that syntax.
.PP
For
.I paddingrule
.TP
.B zero
Pad the irrelevant bits in ASCII-7 and Baudot output with zeros.
.TP
.B printable
Set the values of the irrelevant bits in ASCII-7 and Baudot output to
try to cause the output to stay within character range 32-126. This
option is the default. (Note that
.I paddingrule
is irrelevant for 8-bit character sets like Latin-1 and EBCDIC).
.TP
.B random
Pad the irrelevant bits in ASCII-7 and Baudot output with random data,
except that outputing an all-bits-zero octet is avoided.
.SH CAVEATS
Not all conversions are possible, due to the character sets having
different characters available. In most cases, an unconvertible or
invalid character will be converted to a 0 (padded appropriately); the
exception is that if a tab cannot be converted, it will instead be
converted to a single space (so that INTERCAL programs still run if
converted to Baudot). Using
.B atari
as an input or output character set will literally convert characters
which differ between Atari and Princeton syntax without checking to
see whether they are being used as operators or not.
.PP
If
.I arrayname
is given, then instead of outputting the converted text literally, it
will be output as portable (that is, legal in INTERCAL-72, C-INTERCAL,
J-INTERCAL and CLC-INTERCAL) INTERCAL that dimensions the tail array
given as
.I arrayname
and assigns each byte of the output to an element of that array; this
is mostly useful for producing Baudot text for CLC-INTERCAL-style
array IO.
.SH AUTHOR
.B convickt
and this manual page were written by Alex Smith.


NEWS
			NEW IN THIS RELEASE

	Release ? (by ?) ?

(All these changes were by AIS)

1. Added support for computed ABSTAIN and TRY AGAIN.

2. Added support for computed COME FROM.

3. Added support for multithread programs. This includes ONCE and AGAIN,
which also work in singlethread programs.

4. Expanded intercal.el to support Font Lock and some utilities for generating
INTERCAL-style constants. It also works with the compilation routines of Emacs
to compile INTERCAL programs without the need for make. Degenerated C code now
has a mode line to help with Emacs compatibility.

5. Added a rudimentary debugger and profiler for singlethread programs. The
debugger's features include useful things like 'write out INTERCAL expression
in simpler terms' (which is only possible if -O is on). The profiler works
much like gcov, counting the length of time spent in and number of executions
of each command in the code.

6. Changed searching for includes, libraries, syslib, skeleton. This is so that
non-root users, or people who want to evaluate ick, can do so more simply. The
compiler will search for the above files in the directory they should have been
installed in, the current directory, and the directory in which ick's
executable resides (in that order). Skeleton problems, in particular, seem to
have been a major problem for unfamiliar users trying out ick.

7. Added several new test programs to the Pit.

8. Greatly expanded the original optimizer (which optimizes expressions). It
can now recognize most of the idioms in syslib, and many idioms in other Pit
programs. Also added a new optimizer, to try to optimize program flow, which is
important because otherwise INTERCAL wouldn't have any rudimentary optimizers.
There is also the -F optimizer, which I hope will become infamous. It creates
object code which is close to maximally fast in some specialized situations, at
the expense of compile time (compilation can take a very long time when -F is
active) and object file size. I suggest checking its effect on primes.i.

9. Fixed the grammar (again) in a spark/ears situation involving array
subscripts, after proving that the previous fixes weren't adequate.

10. Added support for backtracking (MAYBE, GO AHEAD, GO BACK). Note that it is
possible for multithreading, backtracking, computed ABSTAIN, and computed COME
FROM to all be used simultaneously in the same program, to help maximize
obfuscation.

11. Added an option (-l) for generating warnings. To prevent this actually being
useful, it is designed to generate far too many warnings, which all have to be
trawled through to find the important ones, and has no way to turn the
unimportant ones off.

12. Added the +printflow and +mystery options for executing INTERCAL programs.
Note that it is not entirely obvious what either of them do (+printflow is a
debugging aid, but the output is somewhat obfuscated, and I am not going to
explain what +mystery does in this document).

	Release 0.24  (by Eric S. Raymond) 31 Dec 2003

1. RPMs can now be built by non-root users.

	Release 0.23  (by Eric S. Raymond) 9 Mar 2003

1. More fixes by Don Knuth.  He found a bug in the type-propagation logic.

	Release 0.22  (by Eric S. Raymond) 4 Mar 2003

1. Incorporated a minor fix patch from Donald Knuth.  Yes, *the*
Donald Knuth.  Seems he found a weakness in the code for handling
comments, and fixed it to do multiline comments properly.  He also made
some improvements to intercal.el, and contributed a test program.

2. Updated build machinery.  This package now uses GNU autoconf.

	Release 0.21 (by Eric S. Raymond) 30 July 2002

1. Jonathan Nieder sent a grammar fix for BLR's old bug #5 (a bad spark-ears
interaction.)

2. Latin-1 support and makefile fixes from Magnus Bodin.

	Release 0.20 (by Eric S. Raymond) 16 September 1999

  (The following changes are due to John Cowan <cowan@ccil.org>)

1.  In addition to Latin-1, UTF-8 is now acceptable in source code as well.
All the numerous currency symbols of Unicode (ISO 10646) can be used as
big-money alternatives in their UTF-8 format.  Indeed, a random mixture of
Latin-1 and UTF-8 also works fine; this seems very much in the spirit of
Intercal.
 
2.  Intercal has now been localized for the Ancient Roman as well as the
Posix/C locale.  You may input numbers in Latin and write Intercal programs
using Latin keywords.  See lexer.l for a list of equivalences.

3.  The Makefile in the pit directory now works like a real Makefile, and
won't compile anything that's already been compiled.
 
	Release 0.19 (by Eric S. Raymond) 16 August 1998

1. Fixed a code-generation error in assignments from tail and hybrid
variables. (Error pointed out by Malcom Ryan.)

	Release 0.18 (by Eric S. Raymond) 29 April 1998

(The following changes are due to John Cowan <cowan@ccil.org>)

1.  The source code is now POSIXly correct.  By making the POSIX
variable in the Makefile undefined, you can restore the use of
<varargs.h>, but otherwise <stdarg.h> is now used.

2.  Latin-1 (ISO 8859-1) is now the standard input charset of C-INTERCAL.
This means that the change operator (0xA2) has returned as an alternative to
big money ($).  Furthermore, the quid (0xA3) and zlotnik (0xA4) are also
acceptable.  For those using Latin-9 (ISO 8859-15), the euro (0xA4) will also
work.  ("Euro", like "ampersand", cannot be sillified further.)
New names have been added to intercal.mm.

3. A fortiori, input of Volap\"uk digits may now be in Latin-1 or TeX format.

4. Yacc is now deprecated in favor of bison; some old yaccs don't do proper
error recovery, and lines like "DO ~ ERROR" failed at compile time.

5.  Some minor portability bugs on systems where ptr_diff is not the same
as int were removed by casting expressions of the form (ptr - ptr) to int.

	Release 0.17 (by Eric S. Raymond) 27 Feb 1998

Incorporates a minor bug fix by Jeff Uphoff <juphoff@tarsier.cv.nrao.edu>.

	Release 0.16 (by Brian Raiter and Eric S. Raymond) 26 May 1997


src/convickt.c
/*****************************************************************************

NAME
    convickt.c -- translate between various INTERCAL formats

LICENSE TERMS
    Copyright (C) 2007 Alex Smith

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

***************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include <unistd.h>

int ick_clc_cset_convert(char* in, /*@partial@*/ char* out, char* incset,
			 char* outcset, int padstyle, size_t outsize,
			 /*@null@*/ FILE* errsto);

/* For communication with clc_cset.c */
/*@observer@*/ char* ick_globalargv0;
/*@-redef@*/ /* it's never the case that both definitions are used at once */
/*@observer@*/ /*@dependent@*/ char* ick_datadir;
/*@=redef@*/
/* We want to read the latest versions of the character sets from disk. */
/*@null@*/ char* ick_clc_cset_atari=0;
/*@null@*/ char* ick_clc_cset_baudot=0;
/*@null@*/ char* ick_clc_cset_ebcdic=0;
/*@null@*/ char* ick_clc_cset_latin1=0;

#ifndef ICKDATADIR
#define ICKDATADIR "/usr/local/share/ick-0.27"
#endif

/*@-redef@*/ /* because only one main is used at a time */
int main(int argc, char** argv)
/*@=redef@*/
{
  size_t allocsize=100;
  char* in;
  char* out;
  size_t i=0;
  int ti;
  int c;
  int padding=-1;
  ick_globalargv0=argv[0];
  srand((unsigned)time(NULL));
  if(argc<3||argc>5)
  {
    fprintf(stderr,"Usage: convickt informat outformat [padding [arrayname]]\n");
    fprintf(stderr,"Available formats are:\n");
    fprintf(stderr,"\tPrinceton: latin1, baudot, ebcdic\n");
    fprintf(stderr,"\tAtari: atari (7-bit ASCII)\n");
    fprintf(stderr,"Possible values for padding are:\n");
    fprintf(stderr,"\tprintable: try to leave the output in char range "
	    "32-126\n");
    fprintf(stderr,"\tzero: leave the output with 0s in the high bits\n");
    fprintf(stderr,"\trandom: pad with random data\n");
    fprintf(stderr,"Padding only has an effect on character sets with less\n");
    fprintf(stderr,"than 8 bits used per bytes; default is 'printable'.\n");
    fprintf(stderr,"If arrayname is given, which must be a legal name for\n");
    fprintf(stderr,"a tail array, the output will instead be a portable\n");
    fprintf(stderr,"INTERCAL program that stores the required byte pattern\n");
    fprintf(stderr,"in that array.\n");
    return 0;
  }
  if(argc>=4&&!strcmp(argv[3],"printable")) padding=1;
  if(argc>=4&&!strcmp(argv[3],"zero")) padding=0;
  if(argc>=4&&!strcmp(argv[3],"random")) padding=2;
  if(argc>=4&&padding==-1)
  {
    fprintf(stderr,"Error: padding value not recognized.\n");
    return 1;
  }
  if(!(ick_datadir=getenv("ICKDATADIR")))
    ick_datadir=ICKDATADIR;
  in=malloc(allocsize);
  if(!in) {perror("Error: Memory allocation failure"); return 0;}
  while((c=getchar())!=EOF)
  {
    in[i++]=(char)c;
    if(i+1>=allocsize)
    {
      char* temp=in;
      allocsize*=2;
      in=realloc(in,allocsize);
      if(!in)
      {
	perror("Error: Memory allocation failure");
	/* Annotation, because in hasn't been freed here; that's what the
	   error return from realloc means. */
	/*@-usereleased@*/
	if(temp) free(temp);
	/*@=usereleased@*/
	return 0;
      }
    }
  }
  in[i]='\0';
  out=malloc(allocsize*6+6);
  if(!out)
  {
    perror("Error: Memory allocation failure");
    free(in);
    return 0;
  }
  ti=ick_clc_cset_convert(in,out,argv[1],argv[2],padding,allocsize*6+6,stderr);
  if(ti>=0&&argc<5) (void) fwrite(out,1,(size_t)ti,stdout);
  else if(ti>=0&&argc==5)
  {
    int pleasedelay=2;
    printf("\tDO %s <- #%d\n",argv[4],ti);
    while(ti--)
    {
      printf("\t");
      if(!--pleasedelay) {printf("PLEASE "); pleasedelay=4;}
      printf("DO %s SUB #%d <- #%d\n",argv[4],ti+1,(int)(out[ti]));
    }
  }
  free(in);
  free(out);
  return ti<0;
}
